torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py

```diff
@@ -65,3 +67,78 @@ def decompress(fut):
         return [decompressed_tensor]
 
     return fut.then(decompress)
+
+
+class AllReduceFirstKStepsState(object):
+    """
+    State information required by allreduce_first_k_steps_hook to run gradient averaging for the first K steps.
+    Need to specify ``k``.
+    """
+
+    __slots__ = [
+        "process_group",
+        "k",
+        "step",
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        k,
+    ):
+        self.process_group = process_group
+        self.k = k
+        self.step = 0
+
+        logging.info(
+            "Allreduce-first-k-steps hook runs allreduce for the first {} steps.".format(
+                self.k,
+            )
+        )
+
+    def maybe_increase_step(self, bucket):
+        # Since bucket 0 is the last bucket to allreduce in an iteration.
+        # Only increase `step` when bucket 0 is processed.
+        if bucket.get_index() == 0:
+            self.step += 1
+
+        if self.step == self.k:
+            logging.info("No more gradient averaging after {} steps.".format(self.k))
+
+
+def allreduce_first_k_steps_hook(
+    state: AllReduceFirstKStepsState, bucket: dist._GradBucket
+) -> torch.futures.Future:
+    """
+    This DDP communication hook runs allreduce to average gradients for the first K steps,
+    and then stops any communication afterwards.
+    This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage.
+    It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.
+
+    Args:
+        state (AllReduceFirstKStepsState): State information to run allreduce for the first K steps.
+        bucket (dist._GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
+            Note that since DDP comm hook only supports single process single device mode at this time,
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
```

Comment:
nit: mention that this will just a completed future with input tensors (local grads) in the case step > k?

Reply:
Good catch!